### PR TITLE
Add `load_if_exists` option

### PR DESCRIPTION
**Pull request by sile**
I added `exist_ok` option to` optuna.study.create_study() `function and `$ optuna create-study` CLI command. It allows users to create studies without warrying about "whether the study already exists in the storage or not".

### CODE REVIEW

File: optuna/cli.py [hunk 1]
```diff
@@ -53,6 +53,10 @@ def get_parser(self, prog_name):
                             default='minimize',
                             help='Set direction of optimization to a new study. Set \'minimize\' '
                                  'for minimization and \'maximize\' for maximization.')
+        parser.add_argument('--exist-ok', default=False, action='store_true',
+                            help='If specified, the creation of the study is skipped '
+                                 'without any error even if the same name study already '
```
Comment (toshihikoyanase) [inline optuna/cli.py]: This message has two if clauses, and it seems difficult to parse. Could you rephrase this sentence?

For instance, I think we can use `when` in this case as follows:
`If specified, the creation of the study is skipped without any errors when the study name is duplicated.`

File: optuna/study.py [hunk 2]
```diff
@@ -472,14 +473,34 @@ def create_study(
         direction:
             Direction of optimization. Set ``minimize`` for minimization and ``maximize`` for
             maximization. Note that ``maximize`` is currently unsupported.
+        exist_ok:
+            Flag for controlling the behavior when study names conflict.
```
Comment (toshihikoyanase) [inline optuna/study.py]: Personally, `to control` seems to be a more common expression than `for controlling`.
I think we can rephrase the sentence as follows:
`Flag to control the behavior to handle a conflict of study names.`

File: optuna/study.py [hunk 3]
```diff
@@ -472,14 +473,34 @@ def create_study(
...
+            If a study named ``study_name`` already exists in the ``storage``,
+            a :class:`~optuna.structs.DuplicatedStudyError` is raised if ``exist_ok`` is
+            set to :obj:`False`.
```
Comment (toshihikoyanase) [inline optuna/study.py]: Similarly to the L57, this message has two if clauses. Please think of a different expression.

File: optuna/study.py [hunk 4]
```diff
@@ -472,14 +473,34 @@ def create_study(
...
 
     """
 
     storage = storages.get_storage(storage)
-    study_name = storage.get_study_name_from_id(storage.create_new_study_id(study_name))
+    try:
+        study_id = storage.create_new_study_id(study_name)
+    except structs.DuplicatedStudyError:
+        if exist_ok:
+            assert study_name is not None
```
Comment (toshihikoyanase) [inline optuna/study.py]: Do we need this assertion? I think we can check the scenario in unit tests.

File: optuna/study.py [hunk 5]
```diff
@@ -472,14 +473,34 @@ def create_study(
...
+
+            logger = logging.get_logger(__name__)
+            logger.info('The same name study exists in the storage. '
```
Comment (toshihikoyanase) [inline optuna/study.py]: `storage` should be hidden from users if possible. I think we can rephrase this sentence as follows:

`'The study name {} is duplicated with the existing study.'.format(study_name)`

File: tests/test_study.py [hunk 6]
```diff
@@ -437,3 +438,77 @@ def f(trial):
...
+
+@pytest.mark.parametrize('storage_mode', STORAGE_MODES)
+def test_create_study(storage_mode):
+    # type: (str) -> None
+
+    with StorageSupplier(storage_mode) as storage:
+        # Test creating a new study.
+        study = optuna.create_study(storage=storage, exist_ok=False)
+
+        if not isinstance(study.storage, optuna.storages.InMemoryStorage):
```
Comment (toshihikoyanase) [inline tests/test_study.py]: Can we apply the early-return pattern here?
And I think the test case with `exist_ok=True` works with InMemoryStorage.

File: tests/test_study.py [hunk 7]
```diff
@@ -437,3 +438,77 @@ def f(trial):
...
+            # Test exist_ok=True with existing study.
+            optuna.create_study(study_name=study.study_name, storage=storage, exist_ok=True)
+
+            # Test exist_ok=False with existing study.
+            with pytest.raises(optuna.structs.DuplicatedStudyError):
+                optuna.create_study(study_name=study.study_name, storage=storage, exist_ok=False)
+
+
+def test_create_study_parallel():
```
Comment (toshihikoyanase) [inline tests/test_study.py]: Does this unit test assume simultaneous creation of study? If so, CPython's GIL will prevent the concurrent execution of threads.

File: optuna/study.py (hunk 4 above)
Comment (sile) [reply optuna/study.py]: If we delete this line and `study_name = None` is passed,  the execution of the code will end with an error raised by https://github.com/pfnet/optuna/blob/0448f096bd6c40c0089f013d90131fad887cd1e3/optuna/storages/rdb/models.py#L52.

Because the deletion of the line does not cause a critical problem (e.g., data inconsistency), I think we can delete it safely.

But I copied this assertion from https://github.com/pfnet/optuna/pull/216/files#diff-4e020643693eb296936f5dd7194bd6d6R397, so I would like to hear @g-votte 's opinion.

File: tests/test_study.py (hunk 6 above)
Comment (sile) [reply tests/test_study.py]: You are right. I'll fix it.

File: tests/test_study.py (hunk 7 above)
Comment (sile) [reply tests/test_study.py]: > Does this unit test assume simultaneous creation of study? 

Yes.
I didn't know about GIL...
I will fix this test case (or maybe I will remove this).

Comment (sile) [reply tests/test_study.py]: I removed this test because the current test is meaningless and testing parallel execution feels to be too costly.

**Comment by sile** [comment]
@toshihikoyanase Thank you for reviewing. I reflected your comments except https://github.com/pfnet/optuna/pull/242#discussion_r238195852.

File: optuna/study.py (hunk 4 above)
Comment (toshihikoyanase) [reply optuna/study.py]: That makes sense. If a name generated automatically is duplicated, this condition (`study_name=None, exist_ok=True`) will be satisfied. In this case, I don't think `ValueError(NOT_FOUND_MSG)` is an appropriate message.
Thank you for your kind explanation.

File: optuna/storages/rdb/storage.py [hunk 8]
```diff
@@ -63,8 +63,9 @@ def create_new_study_id(self, study_name=None):
                                   direction=structs.StudyDirection.NOT_SET)
         session.add(study)
         if not self._commit_with_integrity_check(session):
-            raise ValueError(
-                "study_name {} already exists. Please use a different name.".format(study_name))
+            raise structs.DuplicatedStudyError(
+                "study_name {} already exists. Please use a different name or "
```
Comment (iwiwi) [inline optuna/storages/rdb/storage.py]: How about `Another study with name '{}' already exists.` ?

File: optuna/study.py [hunk 9]
```diff
@@ -472,14 +473,34 @@ def create_study(
         direction:
             Direction of optimization. Set ``minimize`` for minimization and ``maximize`` for
             maximization. Note that ``maximize`` is currently unsupported.
+        exist_ok:
+            Flag to control the behavior to handle a conflict of study names.
+            In the case where a study named ``study_name`` already exists in the ``storage``,
+            a :class:`~optuna.structs.DuplicatedStudyError` is raised if ``exist_ok`` is
+            set to :obj:`False`.
+            Otherwise the creation of the study is skipped and no exception is raised.
```
Comment (iwiwi) [inline optuna/study.py]: We may want to add `and the existing study is returned?`

File: optuna/study.py [hunk 10]
```diff
@@ -472,14 +473,34 @@ def create_study(
...
-    study_name = storage.get_study_name_from_id(storage.create_new_study_id(study_name))
+    try:
+        study_id = storage.create_new_study_id(study_name)
+    except structs.DuplicatedStudyError:
+        if exist_ok:
+            assert study_name is not None
+
+            logger = logging.get_logger(__name__)
+            logger.info('The study name {} is duplicated with the existing study.'
+                        'It is reused instead of creating a new one.'.format(study_name))
```
Comment (iwiwi) [inline optuna/study.py]: How about something like :`Using an existing study with name '{}' instead of creating a new one` or `Connecting to an existing study with name '{}' instead of creating a new one` ?

**Comment by sile** [comment]
@toshihikoyanase @iwiwi 

Thank you for your helpful comments. I reflected those.
And I renamed the option name from `exist_ok` to `load_if_exists` (or `--skip-if-exists` in CLI).

File: optuna/storages/rdb/storage.py [hunk 11]
```diff
@@ -63,8 +63,10 @@ def create_new_study_id(self, study_name=None):
                                   direction=structs.StudyDirection.NOT_SET)
         session.add(study)
         if not self._commit_with_integrity_check(session):
-            raise ValueError(
-                "study_name {} already exists. Please use a different name.".format(study_name))
+            raise structs.DuplicatedStudyError(
+                "Another study with name '{}' already exists. "
+                "Please specify a different name, or reuse the existing one "
+                "by setting `load_if_exists` or `--skip-if-exists` flag.".format(study_name))
```
Comment (g-votte) [inline optuna/storages/rdb/storage.py]: It'd be better to describe that `load_if_exists` is for Python and that `--skip-if-exists` is for CLI. 

```
"Please specify a different name, or reuse the existing one "
"by setting `load_if_exists` (for Python API) or `--skip-if-exists` flag (for CLI).".format(study_name))
```

File: optuna/study.py [hunk 12]
```diff
@@ -472,14 +473,34 @@ def create_study(
...
-    study_name = storage.get_study_name_from_id(storage.create_new_study_id(study_name))
+    try:
+        study_id = storage.create_new_study_id(study_name)
+    except structs.DuplicatedStudyError:
+        if load_if_exists:
+            assert study_name is not None
+
+            logger = logging.get_logger(__name__)
+            logger.info("Using an existing study with name '{}' instead of "
+                        "creating a new one".format(study_name))
```
Comment (g-votte) [inline optuna/study.py]: Please add period at the end of the sentence.

**Comment by sile** [comment]
@g-votte Thank you for your comments. I modified log and error messages for reflecting your review.
